Direct Memory Allocation Failure on Spark 3.1.2 when using arrow

I am using the following versions of spark-snowflake on Spark 3.1.2 ([Databricks Runtime Version 9.1](https://docs.databricks.com/release-notes/runtime/9.1.html)) 

snowflake-ingest-sdk-0.10.2.jar
snowflake-jdbc-3.13.3.jar
spark-snowflake_2.12-2.9.0-spark_3.1.jar

Exception:
```
21/11/17` 21:09:52 INFO SnowflakeTelemetry$: Send OOB Telemetry message: ResultIterator read
21/11/17 21:09:52 ERROR Executor: Exception in task 33.0 in stage 9021.0 (TID 137818)
net.snowflake.client.jdbc.SnowflakeSQLLoggedException: JDBC driver internal error: Max retry reached for the download of #chunk0 (Total chunks: 13) retry=10, error=net.snowflake.client.jdbc.SnowflakeSQLLoggedException: JDBC driver internal error: Exception: Failure allocating buffer..
	at net.snowflake.client.jdbc.SnowflakeChunkDownloader$2.downloadAndParseChunk(SnowflakeChunkDownloader.java:806)
	at net.snowflake.client.jdbc.SnowflakeChunkDownloader$2.call(SnowflakeChunkDownloader.java:861)
	at net.snowflake.client.jdbc.SnowflakeChunkDownloader$2.call(SnowflakeChunkDownloader.java:772)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: net.snowflake.client.jdbc.internal.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PooledByteBufAllocatorL.allocate(PooledByteBufAllocatorL.java:60)
	at net.snowflake.client.jdbc.internal.apache.arrow.memory.AllocationManager.<init>(AllocationManager.java:78)
	at net.snowflake.client.jdbc.internal.apache.arrow.memory.BaseAllocator.bufferWithoutReservation(BaseAllocator.java:347)
	at net.snowflake.client.jdbc.internal.apache.arrow.memory.BaseAllocator.buffer(BaseAllocator.java:315)
	at net.snowflake.client.jdbc.internal.apache.arrow.memory.BaseAllocator.buffer(BaseAllocator.java:279)
	at net.snowflake.client.jdbc.internal.apache.arrow.vector.ipc.message.MessageSerializer.readMessageBody(MessageSerializer.java:695)
	at net.snowflake.client.jdbc.internal.apache.arrow.vector.ipc.message.MessageChannelReader.readNext(MessageChannelReader.java:68)
	at net.snowflake.client.jdbc.internal.apache.arrow.vector.ipc.ArrowStreamReader.loadNextBatch(ArrowStreamReader.java:106)
	at net.snowflake.client.jdbc.ArrowResultChunk.readArrowStream(ArrowResultChunk.java:78)
	at net.snowflake.client.jdbc.SnowflakeChunkDownloader$2.downloadAndParseChunk(SnowflakeChunkDownloader.java:790)
	... 6 more
Caused by: net.snowflake.client.jdbc.internal.io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 43822088192, max: 43822088192)
	at net.snowflake.client.jdbc.internal.io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:640)
	at net.snowflake.client.jdbc.internal.io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:594)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:764)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:740)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:244)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PoolArena.allocate(PoolArena.java:226)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PoolArena.allocate(PoolArena.java:146)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PooledByteBufAllocatorL$InnerAllocator.newDirectBufferL(PooledByteBufAllocatorL.java:173)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PooledByteBufAllocatorL$InnerAllocator.directBuffer(PooledByteBufAllocatorL.java:206)
	at net.snowflake.client.jdbc.internal.io.netty.buffer.PooledByteBufAllocatorL.allocate(PooledByteBufAllocatorL.java:58)
	... 15 more
.
	at net.snowflake.client.jdbc.SnowflakeChunkDownloader.getNextChunkToConsume(SnowflakeChunkDownloader.java:510)
	at net.snowflake.client.core.SFArrowResultSet.fetchNextRowUnsorted(SFArrowResultSet.java:233)
	at net.snowflake.client.core.SFArrowResultSet.fetchNextRow(SFArrowResultSet.java:210)
	at net.snowflake.client.core.SFArrowResultSet.next(SFArrowResultSet.java:346)
	at net.snowflake.client.jdbc.SnowflakeResultSetV1.next(SnowflakeResultSetV1.java:99)
	at net.snowflake.spark.snowflake.io.ResultIterator.hasNext(SnowflakeResultSetRDD.scala:142)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:132)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.$anonfun$runTask$3(ShuffleMapTask.scala:81)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.scheduler.ShuffleMapTask.$anonfun$runTask$1(ShuffleMapTask.scala:81)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:150)
	at org.apache.spark.scheduler.Task.$anonfun$run$1(Task.scala:119)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.scheduler.Task.run(Task.scala:91)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$13(Executor.scala:813)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1605)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:816)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:672)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
21/11/17 21:09:52 INFO CoarseGrainedExecutorBackend: Got assigned task 138546
```

Setting use_copy_unload=true fixes the issue. Any ideas on why snowflake-spark-connector is consuming ~43GB of direct memory? It doesn't seem like an issue on Spark 3.1.1

--- a/src/main/scala/net/snowflake/spark/snowflake/io/SnowflakeResultSetRDD.scala
+++ b/src/main/scala/net/snowflake/spark/snowflake/io/SnowflakeResultSetRDD.scala
@@ -45,6 +45,7 @@ class SnowflakeResultSetRDD[T: ClassTag](
     SparkConnectorContext.recordConfig()
 
     ResultIterator[T](
+      context: TaskContext,
       schema,
       split.asInstanceOf[SnowflakeResultSetPartition].resultSet,
       split.asInstanceOf[SnowflakeResultSetPartition].index,
@@ -62,6 +63,7 @@ class SnowflakeResultSetRDD[T: ClassTag](
 }
 
 case class ResultIterator[T: ClassTag](
+  context: TaskContext,
   schema: StructType,
   resultSet: SnowflakeResultSetSerializable,
   partitionIndex: Int,
@@ -125,9 +127,19 @@ case class ResultIterator[T: ClassTag](
     }
   }
 
+  context.addTaskCompletionListener[Unit](_ => closeResultSet())
+
   val isIR: Boolean = isInternalRow[T]
   val mapper: ObjectMapper = new ObjectMapper()
   var currentRowNotConsumedYet: Boolean = false
+  var resultSetIsClosed: Boolean = false
+
+  private def closeResultSet(): Unit = {
+    if (!resultSetIsClosed) {
+      data.close()
+      resultSetIsClosed = true
+    }
+  }
 
   override def hasNext: Boolean = {
     // In some cases, hasNext() may be called but next() isn't.
@@ -150,8 +162,7 @@ case class ResultIterator[T: ClassTag](
              | actualReadRowCount=$actualReadRowCount
              |""".stripMargin.filter(_ >= ' '))
 
-        // Close the result set.
-        data.close()
+        closeResultSet()
 
         // Inject Exception for test purpose
         TestHook.raiseExceptionIfTestFlagEnabled(

--- a/src/main/scala/net/snowflake/spark/snowflake/io/SnowflakeResultSetRDD.scala
+++ b/src/main/scala/net/snowflake/spark/snowflake/io/SnowflakeResultSetRDD.scala
@@ -45,7 +45,6 @@ class SnowflakeResultSetRDD[T: ClassTag](
     SparkConnectorContext.recordConfig()
 
     ResultIterator[T](
-      context: TaskContext,
       schema,
       split.asInstanceOf[SnowflakeResultSetPartition].resultSet,
       split.asInstanceOf[SnowflakeResultSetPartition].index,
@@ -63,7 +62,6 @@ class SnowflakeResultSetRDD[T: ClassTag](
 }
 
 case class ResultIterator[T: ClassTag](
-  context: TaskContext,
   schema: StructType,
   resultSet: SnowflakeResultSetSerializable,
   partitionIndex: Int,
@@ -127,17 +125,17 @@ case class ResultIterator[T: ClassTag](
     }
   }
 
-  context.addTaskCompletionListener[Unit](_ => closeResultSet())
-
   val isIR: Boolean = isInternalRow[T]
   val mapper: ObjectMapper = new ObjectMapper()
   var currentRowNotConsumedYet: Boolean = false
   var resultSetIsClosed: Boolean = false
 
+  TaskContext.get().addTaskCompletionListener[Unit](_ => closeResultSet())
+
   private def closeResultSet(): Unit = {
     if (!resultSetIsClosed) {
-      data.close()
       resultSetIsClosed = true
+      data.close()
     }
   }
 

